test/common/testlib.py

```diff
@@ -783,18 +783,18 @@ def assert_pixels(self, selector, key, ignore=[]):
 
         rect = self.call_js_func('ph_element_clip', selector)
 
-        def relative_clip(sel):
-            r = self.call_js_func('ph_element_clip', sel)
-            return (r['x'] - rect['x'],
-                    r['y'] - rect['y'],
-                    r['x'] - rect['x'] + r['width'],
-                    r['y'] - rect['y'] + r['height'])
+        def relative_clips(sels):
+            return list(map(lambda r: (r['x'] - rect['x'],
+                                       r['y'] - rect['y'],
+                                       r['x'] - rect['x'] + r['width'],
+                                       r['y'] - rect['y'] + r['height']),
+                            self.call_js_func('ph_selector_clips', sels)))
 
         reference_dir = os.path.join(TEST_DIR, 'reference')
         if not os.path.exists(os.path.join(reference_dir, '.git')):
             subprocess.check_call([f'{TEST_DIR}/common/pixel-tests', 'pull'])
 
-        ignore_rects = list(map(relative_clip, map(lambda item: selector + " " + item, ignore)))
+        ignore_rects = relative_clips(list(map(lambda item: selector + " " + item, ignore)))
```

Comment:
You shouldn't need the `list()` here.

Reply:
A "map iterator" can not be passed to json.dumps, which is used by `self.call_js_func('ph_selector_clips', sels)` to convert the arguments to JS.  We could do `self.call_js_func('ph_selector_clips', list(sels))` but then `relative_clips` would have the side effect of consuming its argument, which I find odd.

Reply:
And just to flog this dead horse some more, consider this:

```
>>> json.dumps([1, 2, 3])
'[1, 2, 3]'
>>> json.dumps(map(lambda x: 2*x, [1, 2, 3]))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib64/python3.10/json/__init__.py", line 231, in dumps
    return _default_encoder.encode(obj)
  File "/usr/lib64/python3.10/json/encoder.py", line 199, in encode
    chunks = self.iterencode(o, _one_shot=True)
  File "/usr/lib64/python3.10/json/encoder.py", line 257, in iterencode
    return _iterencode(o, 0)
  File "/usr/lib64/python3.10/json/encoder.py", line 179, in default
    raise TypeError(f'Object of type {o.__class__.__name__} '
TypeError: Object of type map is not JSON serializable
```

I find this bizarre, from a "Functional Programming" point of view.  I would expect `map` to only return an iterator when mapping over iterators. If this is supposed to emulate lazy evaluation, then an iterator should not be consumed when iterating over it.

Anyway, `list(map(...))` it is from now on, and always! :-)

Reply:
Hehe, thanks for the research!